### DISCUSSION TITLE
Cannot use GCNConv with pytorch.cuda.amp

### DISCUSSION THREAD

**Discussion by dxyzx0** [pyg-team/pytorch_geometric#3036] [Q&A]
![image](https://user-images.githubusercontent.com/28868457/130899721-38d62c57-6c9b-409b-9211-ec1095929267.png)
Cannot use `pytorch.cuda.amp` since some tensor in `GCNConv` doesn't change `dtype` to `half float`
PS: `sgcn_conv.py` is just a rewrite of `gcn_conv.py` with only code modification of `gcn_norm`, I think it won't affect `adj_t`

I find that `SparseTensor` will be `autocast` by `torch.cuda.amp`, that's where the problem resides.
![Uploading image.png…]()

**Comment by rusty1s** [comment]
Does the same code work for you when using the `edge_index` representation instead of `SparseTensor`? And do you have a small script to reproduce? `torch-sparse` supports half-precision by now, and conversion of `adj_t` to `torch.half` should work as well.

**Reply by dxyzx0** [reply]
Maybe that's because I construct my data like this?
![image](https://user-images.githubusercontent.com/28868457/130918907-e04bf666-e3b9-4621-9e44-ff156958a8bb.png)

**Reply by dxyzx0** [reply]
AMP works using the `edge_index` representation. If I comment the 3 lines below `for data in train_loader`, amp will work.

**Reply by rusty1s** [reply]
I can look into it. If you have a minimal example to share, please do so. That would help me out tremendously.

**Reply by dxyzx0** [reply]
```
import torch
import torch_geometric
from torch_geometric.data import Data
from torch_sparse import SparseTensor


class Net(torch.nn.Module):
    def __init__(self, device):
        super(Net, self).__init__()
        self.device = device
        self.gcn = torch_geometric.nn.GCNConv(2, 3)

    def forward(self, data: SparseTensor):
        X, Adj, edge_weight = data.x, data.edge_index, data.edge_attr
        return self.gcn(X, Adj, edge_weight)


if __name__ == "__main__":
    edge_index = torch.LongTensor([[0, 1], [1, 0]])
    edge_weight = torch.Tensor([0.1, 0.2])
    x = torch.Tensor([[0, 1], [1, 1]])
    y = torch.Tensor([0, 1])
    data = Data(x=x, edge_index=edge_index, edge_weight=edge_weight, y=y)

    model = Net(device="cuda")
    model.to(model.device)
    with torch.cuda.amp.autocast():
        data.edge_index = SparseTensor(row=data.edge_index[0], col=data.edge_index[1], value=data.edge_attr,
                                       sparse_sizes=(data.x.size(0), data.x.size(0)))
        data.edge_attr = None
        data = data.to(model.device)

        out = model(data)

```
And what I get:
![image](https://user-images.githubusercontent.com/28868457/130922656-aa6a326c-fa88-42bf-8f27-60f61b5ee1c9.png)

**Reply by dxyzx0** [reply]
My env:
![image](https://user-images.githubusercontent.com/28868457/130925951-c1ee0455-4c98-4350-82cf-68a504bd76ac.png)

**Reply by rusty1s** [reply]
Thanks for reporting. I fixed it in the latest `torch-sparse` commit, see [here](https://github.com/rusty1s/pytorch_sparse/commit/87c88d95e7d7aaf93581108210b22e45b3d386ff).

**Reply by rusty1s** [reply]
It seems that `x.pt` already contains `inf` values. One additional think that might cause this is the `eps` value of `LayerNorm`, which you may want to set to a higher value in a half-precision scenario.
